// @serverless-cd/goat-ui/src/widgets/OverViewImage/src/utils.ts
// import SentinelImage from "./assets/sentinel.png";

export const hoverContentMockData = {
  title: "云原生xxxx",
  image:
    "https://gw.alicdn.com/imgextra/i3/O1CN01dI5UtV1iMnEY2sB3r_!!6000000004399-2-tps-216-154.png",
  labels: ["Spring", "Java", "MQ"],
  links: {
    Github: {
      link: "https://github.com/apache/rocketmq",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://rocketmq.apache.org/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://rocketmq.apache.org/docs/quick-start/",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://rocketmq.apache.org/docs/quick-start/",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://rocketmq.apache.org/docs/quick-start/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverSentinelData = {
  title: "Sentinel是⼀款面向分布式、多语言异构化服务架构的流量治理组件。",
  image:
    "https://img.alicdn.com/imgextra/i4/O1CN01GNcMeH1n8f5u0Qus8_!!6000000005045-55-tps-200-36.svg",
  labels: ["高可用", "流量治理", "流量防护"],
  links: {
    Github: {
      link: "https://github.com/alibaba/Sentinel",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://sentinelguard.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/alibaba/Sentinel/releases/tag/1.8.6",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://sentinelguard.io/zh-cn/docs/quick-start.html",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://start.aliyun.com/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverChaosBladeData = {
  title: "ChaosBlade是⼀个云原生混沌工程平台，支持多种环境、集群和语言。",
  image:
    "https://img.alicdn.com/imgextra/i2/O1CN01uCl9p31NxKuC6WxON_!!6000000001636-55-tps-200-36.svg",
  labels: ["高可用", "云原生", "混沌工程"],
  links: {
    Github: {
      link: "https://github.com/chaosblade-io/chaosblade",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://chaosblade.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/chaosblade-io/chaosblade/releases/tag/v1.7.0",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://chaosblade.io/docs",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverAppActiveData = {
  title: "一款标准、通用且功能强大的，致力于构建应用多活架构的开源中间件。",
  image:
    "https://img.alicdn.com/imgextra/i3/O1CN018Q4Lut1HqZMI4Rgot_!!6000000000809-55-tps-200-43.svg",
  labels: ["多活容灾", "高可用", "稳定性"],
  links: {
    Github: {
      link: "https://github.com/alibaba/Appactive",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://doc.appactive.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/alibaba/Appactive/releases/tag/v0.2.1",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://doc.appactive.io/docs/cn/README_CN.html/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverNacosData = {
  title: "⼀个更易于构建云原生应用的动态服务发现、配置管理和服务管理平台。",
  image:
    "https://img.alicdn.com/imgextra/i1/O1CN01YjDURc26ODF5FQt4d_!!6000000007651-55-tps-123-24.svg",
  labels: ["配置管理", "服务发现", "服务管理"],
  links: {
    Github: {
      link: "https://github.com/alibaba/nacos",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://nacos.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/alibaba/nacos/releases/tag/2.1.2",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://nacos.io/zh-cn/docs/v2/quickstart/quick-start.html",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://start.aliyun.com/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverOpenSergoData = {
  title: "OpenSergo是⼀套开放通用的、面向云原生服务、覆盖微服务及上下游关联组件的微服务治理标准。",
  image:
    "https://img.alicdn.com/imgextra/i2/O1CN01Bs0YtX1qMoAvDQO9j_!!6000000005482-55-tps-200-36.svg",
  labels: ["微服务治理", "治理标准", "流量治理"],
  links: {
    Github: {
      link: "https://github.com/opensergo/opensergo-specification",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://opensergo.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/opensergo/opensergo-java-sdk/releases/tag/v0.1.0-beta1",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://opensergo.io/zh-cn/docs/quick-start/opensergo-control-plane/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverHigressData = {
  title: "⼀个遵循开源Ingress/Gateway API标准，提供流量调度、服务治理、安全防护三合⼀的高集成、易使用、易扩展、热更新的下⼀代云原生网关。",
  image:
    "https://img.alicdn.com/imgextra/i2/O1CN0186oyN31ksVtlINQGM_!!6000000004739-55-tps-200-36.svg",
  labels: ["流量网关", "微服务网关", "安全网关"],
  links: {
    Github: {
      link: "https://github.com/alibaba/higress",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://higress.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/alibaba/higress/releases/tag/v0.5.0",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://higress.io/zh-cn/docs/overview/what-is-higress.html",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverRocketMQData = {
  title: "云原生“消息、事件、流”实时数据处理平台，覆盖云边端⼀体化数据处理场景。",
  image:
    "https://img.alicdn.com/imgextra/i1/O1CN01vJbfLg1Vtgfyuv3KF_!!6000000002711-2-tps-447-161.png",
  labels: ["消息", "事件", "流"],
  links: {
    Github: {
      link: "https://github.com/apache/rocketmq",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://rocketmq.apache.org/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/apache/rocketmq/releases/tag/rocketmq-all-5.0.0",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://rocketmq.apache.org/zh/docs/quickStart/02quickstart",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://start.aliyun.com/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverDubboData = {
  title: "Apache Dubbo是⼀款微服务框架，为⼤规模微服务 实践提供⾼性能RPC通信、流量治理、可观测性等解决⽅案，涵盖Java、Golang等多种语⾔SDK实现。",
  image:
    "https://img.alicdn.com/imgextra/i3/O1CN01Efw9Qu1Xto0ghUDgv_!!6000000002982-55-tps-200-36.svg",
  labels: ["微服务", "服务调用", "RPC通信"],
  links: {
    Github: {
      link: "https://github.com/apache/dubbo",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://dubbo.apache.org/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/apache/dubbo/releases/tag/dubbo-3.1.3",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://cn.dubbo.apache.org/zh-cn/overview/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverSCAData = {
  title: "一站式的分布式应用开发框架。",
  image:
    "https://img.alicdn.com/imgextra/i4/O1CN01GtWTyl1aLP6ky10ZS_!!6000000003313-2-tps-1212-114.png",
  labels: ["应用框架", "微服务"],
  links: {
    Github: {
      link: "https://github.com/alibaba/spring-cloud-alibaba",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://spring.io/projects/spring-cloud-alibaba",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/alibaba/spring-cloud-alibaba/releases/tag/2021.0.4.0",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://github.com/alibaba/spring-cloud-alibaba/blob/2021.x/README-zh.md",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://start.aliyun.com/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverSeataData = {
  title: "Seata是⼀款开源的分布式事务解决方案，致力于在微服务架构下提供高性能和简单易用的分布式事务服务。",
  image:
    "https://img.alicdn.com/imgextra/i3/O1CN01ULY29M1JbZchZLegx_!!6000000001047-55-tps-200-36.svg",
  labels: ["分布式事务", "数据一致性", "微服务"],
  links: {
    Github: {
      link: "https://github.com/seata/seata",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://seata.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/seata/seata/releases/tag/v1.5.2",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://seata.io/zh-cn/docs/ops/deploy-guide-beginner.html",
      linkClassname: "text-secondary text-sm",
    },
    快速构建: {
      link: "https://start.aliyun.com/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverKubeVelaData = {
  title: "是⼀个现代化的软件交付平台，它可以让你的应用交付在当今流行的混合、多云环境中变得更加简单、高效、可靠。",
  image:
    "https://img.alicdn.com/imgextra/i4/O1CN010jG1R61qE6eHJs89C_!!6000000005463-55-tps-200-36.svg",
  labels: ["软件交付", "应用管理", "多云部署"],
  links: {
    Github: {
      link: "https://github.com/kubevela/kubevela",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://kubevela.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/kubevela/kubevela/releases/tag/v1.6.4",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://kubevela.io/zh/docs/quick-start",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverKubernetesData = {
  title: "Kubernetes是⼀个开源的容器编排引擎，用来对容器化应用进行自动化部署、扩缩和管理。",
  image:
    "https://img.alicdn.com/imgextra/i3/O1CN012u7Nuk1fQ3CjgvLyT_!!6000000004000-55-tps-201-37.svg",
  labels: ["容器编排", "应用管理", "发布部署"],
  links: {
    Github: {
      link: "https://github.com/kubernetes/kubernetes",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://kubernetes.io/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/kubernetes/kubernetes/releases/tag/v1.23.14",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://kubernetes.io/zh-cn/docs/setup/",
      linkClassname: "text-secondary text-sm",
    },
  },
};

export const hoverSkyWalkingData = {
  title: "⼀款用于分布式系统的应用程序性能监视工具，尤其是面向微服务、云原生和基于容器(Kubernetes)架构设计。",
  image:
    "https://img.alicdn.com/imgextra/i1/O1CN014YV4JL1ZfGfcEh1ur_!!6000000003221-55-tps-200-36.svg",
  labels: ["链路追踪", "应用诊断", "性能监控"],
  links: {
    Github: {
      link: "https://github.com/apache/skywalking",
      stars: 20407,
      forks: 11383,
      titleClassName: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    官网: {
      link: "https://skywalking.apache.org/",
      titleStyle: "text-gray-500",
      linkClassname: "text-secondary text-sm",
    },
    最新版本: {
      link: "https://github.com/apache/skywalking/releases/tag/v9.3.0",
      linkClassname: "text-secondary text-sm",
    },
    快速入门: {
      link: "https://skywalking.apache.org/docs/skywalking-showcase/latest/readme/",
      linkClassname: "text-secondary text-sm",
    },
  },
};
